File: movie_catalog.rb
# movie_catalog.rb
require 'json'
require 'date'

class Movie
  attr_accessor :id, :title, :director, :year, :genre, :watched, :rating, :added_date

  def initialize(id, title, director, year, genre, watched, rating, added_date = Date.today.to_s)
    @id = id
    @title = title
    @director = director
    @year = year
    @genre = genre
    @watched = watched
    @rating = rating
    @added_date = added_date
  end

  def to_h
    { id: @id, title: @title, director: @director, year: @year, genre: @genre,
      watched: @watched, rating: @rating, added_date: @added_date }
  end

  def self.from_h(hash)
    Movie.new(hash[:id], hash[:title], hash[:director], hash[:year],
              hash[:genre], hash[:watched], hash[:rating], hash[:added_date])
  end
end

class MovieCatalog
  attr_reader :movies

  def initialize
    @movies = []
    @next_id = 1
  end

  def add_movie(title, director, year, genre, watched, rating)
    raise "Оценка должна быть от 1 до 10" unless (1..10).include?(rating)
    raise "Год должен быть от 1900 до #{Date.today.year}" unless (1900..Date.today.year).include?(year)
    raise "Название и режиссёр не могут быть пустыми" if title.empty? || director.empty?
    genre = "Неизвестно" if genre.empty?
    movie = Movie.new(@next_id, title, director, year, genre, watched, rating)
    @movies << movie
    @next_id += 1
    movie
  end

  def find_movie(id)
    @movies.find { |m| m.id == id }
  end

  def edit_movie(id, **kwargs)
    movie = find_movie(id)
    return false unless movie
    kwargs.each do |key, value|
      movie.send("#{key}=", value) if movie.respond_to?("#{key}=")
    end
    true
  end

  def delete_movie(id)
    movie = find_movie(id)
    return false unless movie
    @movies.delete(movie)
    true
  end

  def search_movies(query)
    q = query.downcase
    @movies.select { |m| m.title.downcase.include?(q) || m.director.downcase.include?(q) }
  end

  def filter_by_watched(watched)
    @movies.select { |m| m.watched == watched }
  end

  def filter_by_genre(genre)
    @movies.select { |m| m.genre.downcase == genre.downcase }
  end

  def filter_by_year(year)
    @movies.select { |m| m.year == year }
  end

  def stats
    total = @movies.size
    watched_count = filter_by_watched(true).size
    unwatched = total - watched_count
    watched_ratings = filter_by_watched(true).map(&:rating)
    avg_rating = watched_ratings.empty? ? 0 : watched_ratings.sum.to_f / watched_ratings.size
    genres = Hash.new(0)
    @movies.each { |m| genres[m.genre] += 1 }
    { total: total, watched: watched_count, unwatched: unwatched, avg_rating: avg_rating, genres: genres }
  end

  def save_to_file(filename = "movies_data.json")
    data = { movies: @movies.map(&:to_h) }
    File.write(filename, JSON.pretty_generate(data))
  end

  def load_from_file(filename = "movies_data.json")
    return unless File.exist?(filename)
    data = JSON.parse(File.read(filename), symbolize_names: true)
    @movies.clear
    data[:movies].each do |item|
      movie = Movie.from_h(item)
      @movies << movie
      @next_id = movie.id + 1 if movie.id >= @next_id
    end
  rescue JSON::ParserError
    puts "Ошибка чтения файла."
  end
end

def print_movie(movie)
  status = movie.watched ? "✅ Просмотрен" : "⏳ Не просмотрен"
  puts "##{movie.id} - #{movie.title} (#{movie.year})"
  puts "   Режиссёр: #{movie.director}, Жанр: #{movie.genre}"
  puts "   #{status}, Оценка: #{movie.rating}/10"
  puts "   Добавлен: #{movie.added_date}"
end

def main
  catalog = MovieCatalog.new
  catalog.load_from_file

  loop do
    puts "\n===== КАТАЛОГ ФИЛЬМОВ (Ruby) ====="
    puts "1. Добавить фильм"
    puts "2. Показать все фильмы"
    puts "3. Показать просмотренные фильмы"
    puts "4. Показать непросмотренные фильмы"
    puts "5. Найти фильмы по названию или режиссёру"
    puts "6. Отметить фильм как просмотренный"
    puts "7. Редактировать фильм"
    puts "8. Удалить фильм"
    puts "9. Показать статистику"
    puts "10. Сохранить в файл"
    puts "11. Загрузить из файла"
    puts "0. Выход"
    print "Выберите действие: "
    choice = gets.chomp

    case choice
    when "0"
      break
    when "1"
      print "Название: "
      title = gets.chomp
      next if title.empty?
      print "Режиссёр: "
      director = gets.chomp
      next if director.empty?
      print "Год выпуска: "
      year = gets.chomp.to_i
      print "Жанр: "
      genre = gets.chomp
      print "Статус (1-просмотрен, 0-не просмотрен): "
      watched = gets.chomp == "1"
      print "Оценка (1-10): "
      rating = gets.chomp.to_i
      begin
        movie = catalog.add_movie(title, director, year, genre, watched, rating)
        puts "Фильм добавлен с ID #{movie.id}"
      rescue => e
        puts "Ошибка: #{e.message}"
      end
    when "2"
      if catalog.movies.empty?
        puts "Нет фильмов."
      else
        catalog.movies.each { |m| print_movie(m) }
      end
    when "3"
      movies = catalog.filter_by_watched(true)
      if movies.empty?
        puts "Нет просмотренных фильмов."
      else
        movies.each { |m| print_movie(m) }
      end
    when "4"
      movies = catalog.filter_by_watched(false)
      if movies.empty?
        puts "Нет непросмотренных фильмов."
      else
        movies.each { |m| print_movie(m) }
      end
    when "5"
      print "Введите часть названия или режиссёра: "
      query = gets.chomp
      results = catalog.search_movies(query)
      if results.empty?
        puts "Фильмы не найдены."
      else
        results.each { |m| print_movie(m) }
      end
    when "6"
      print "Введите ID фильма: "
      id = gets.chomp.to_i
      movie = catalog.find_movie(id)
      unless movie
        puts "Фильм не найден."
        next
      end
      movie.watched = true
      puts "Фильм отмечен как просмотренный."
    when "7"
      print "Введите ID фильма для редактирования: "
      id = gets.chomp.to_i
      movie = catalog.find_movie(id)
      unless movie
        puts "Фильм не найден."
        next
      end
      puts "Оставьте поле пустым, чтобы не менять."
      print "Название (#{movie.title}): "
      new_title = gets.chomp
      print "Режиссёр (#{movie.director}): "
      new_director = gets.chomp
      print "Год (#{movie.year}): "
      new_year = gets.chomp
      print "Жанр (#{movie.genre}): "
      new_genre = gets.chomp
      print "Статус (1-просмотрен, 0-не просмотрен) сейчас: #{movie.watched ? '1' : '0'}: "
      new_watched = gets.chomp
      print "Оценка (#{movie.rating}): "
      new_rating = gets.chomp
      updates = {}
      updates[:title] = new_title unless new_title.empty?
      updates[:director] = new_director unless new_director.empty?
      unless new_year.empty?
        updates[:year] = new_year.to_i
      end
      updates[:genre] = new_genre unless new_genre.empty?
      unless new_watched.empty?
        updates[:watched] = new_watched == "1"
      end
      unless new_rating.empty?
        updates[:rating] = new_rating.to_i
      end
      if catalog.edit_movie(id, **updates)
        puts "Фильм обновлён."
      else
        puts "Ошибка обновления."
      end
    when "8"
      print "Введите ID фильма для удаления: "
      id = gets.chomp.to_i
      if catalog.delete_movie(id)
        puts "Фильм удалён."
      else
        puts "Фильм не найден."
      end
    when "9"
      stats = catalog.stats
      puts "\n=== СТАТИСТИКА ==="
      puts "Всего фильмов: #{stats[:total]}"
      puts "Просмотрено: #{stats[:watched]}"
      puts "Не просмотрено: #{stats[:unwatched]}"
      puts "Средняя оценка (только просмотренные): #{'%.2f' % stats[:avg_rating]}"
      puts "По жанрам:"
      stats[:genres].each { |g, c| puts "  #{g}: #{c}" }
    when "10"
      catalog.save_to_file
      puts "Сохранено."
    when "11"
      catalog.load_from_file
      puts "Загружено."
    else
      puts "Неизвестная команда."
    end
  end
end

main if __FILE__ == $0

File: movie_catalog.py
# movie_catalog.py
import json
from dataclasses import dataclass, asdict
from datetime import date
from typing import List, Optional
from pathlib import Path

@dataclass
class Movie:
    id: int
    title: str
    director: str
    year: int
    genre: str
    watched: bool
    rating: int  # 1-10
    added_date: str

class MovieCatalog:
    def __init__(self):
        self.movies: List[Movie] = []
        self.next_id = 1

    def add_movie(self, title: str, director: str, year: int, genre: str, watched: bool, rating: int) -> Movie:
        if rating < 1 or rating > 10:
            raise ValueError("Оценка должна быть от 1 до 10")
        if year < 1900 or year > date.today().year:
            raise ValueError(f"Год должен быть от 1900 до {date.today().year}")
        if not title or not director:
            raise ValueError("Название и режиссёр не могут быть пустыми")
        movie = Movie(
            id=self.next_id,
            title=title,
            director=director,
            year=year,
            genre=genre or "Неизвестно",
            watched=watched,
            rating=rating,
            added_date=date.today().isoformat()
        )
        self.movies.append(movie)
        self.next_id += 1
        return movie

    def find_movie(self, movie_id: int) -> Optional[Movie]:
        return next((m for m in self.movies if m.id == movie_id), None)

    def edit_movie(self, movie_id: int, **kwargs) -> bool:
        movie = self.find_movie(movie_id)
        if not movie:
            return False
        for key, value in kwargs.items():
            if hasattr(movie, key) and value is not None:
                setattr(movie, key, value)
        return True

    def delete_movie(self, movie_id: int) -> bool:
        movie = self.find_movie(movie_id)
        if movie:
            self.movies.remove(movie)
            return True
        return False

    def search_movies(self, query: str) -> List[Movie]:
        q = query.lower()
        return [m for m in self.movies if q in m.title.lower() or q in m.director.lower()]

    def filter_by_watched(self, watched: bool) -> List[Movie]:
        return [m for m in self.movies if m.watched == watched]

    def filter_by_genre(self, genre: str) -> List[Movie]:
        return [m for m in self.movies if m.genre.lower() == genre.lower()]

    def filter_by_year(self, year: int) -> List[Movie]:
        return [m for m in self.movies if m.year == year]

    def get_stats(self) -> dict:
        total = len(self.movies)
        watched_count = len(self.filter_by_watched(True))
        unwatched = total - watched_count
        watched_ratings = [m.rating for m in self.movies if m.watched]
        avg_rating = sum(watched_ratings) / len(watched_ratings) if watched_ratings else 0.0
        genres = {}
        for m in self.movies:
            genres[m.genre] = genres.get(m.genre, 0) + 1
        return {
            "total": total,
            "watched": watched_count,
            "unwatched": unwatched,
            "avg_rating": avg_rating,
            "genres": genres
        }

    def save_to_file(self, filename: str = "movies_data.json") -> None:
        data = {"movies": [asdict(m) for m in self.movies]}
        with open(filename, "w", encoding="utf-8") as f:
            json.dump(data, f, ensure_ascii=False, indent=2)

    def load_from_file(self, filename: str = "movies_data.json") -> None:
        path = Path(filename)
        if not path.exists():
            return
        with open(filename, "r", encoding="utf-8") as f:
            data = json.load(f)
            self.movies.clear()
            for item in data.get("movies", []):
                movie = Movie(
                    id=item["id"],
                    title=item["title"],
                    director=item["director"],
                    year=item["year"],
                    genre=item["genre"],
                    watched=item["watched"],
                    rating=item["rating"],
                    added_date=item["added_date"]
                )
                self.movies.append(movie)
                if movie.id >= self.next_id:
                    self.next_id = movie.id + 1

def print_movie(movie: Movie) -> None:
    status = "✅ Просмотрен" if movie.watched else "⏳ Не просмотрен"
    print(f"#{movie.id} - {movie.title} ({movie.year})")
    print(f"   Режиссёр: {movie.director}, Жанр: {movie.genre}")
    print(f"   {status}, Оценка: {movie.rating}/10")
    print(f"   Добавлен: {movie.added_date}")

def main():
    catalog = MovieCatalog()
    catalog.load_from_file()

    while True:
        print("\n===== КАТАЛОГ ФИЛЬМОВ (Python) =====")
        print("1. Добавить фильм")
        print("2. Показать все фильмы")
        print("3. Показать просмотренные фильмы")
        print("4. Показать непросмотренные фильмы")
        print("5. Найти фильмы по названию или режиссёру")
        print("6. Отметить фильм как просмотренный")
        print("7. Редактировать фильм")
        print("8. Удалить фильм")
        print("9. Показать статистику")
        print("10. Сохранить в файл")
        print("11. Загрузить из файла")
        print("0. Выход")
        choice = input("Выберите действие: ").strip()

        if choice == "0":
            break
        elif choice == "1":
            title = input("Название: ").strip()
            if not title:
                print("Название не может быть пустым.")
                continue
            director = input("Режиссёр: ").strip()
            if not director:
                print("Режиссёр не может быть пустым.")
                continue
            try:
                year = int(input("Год выпуска: ").strip())
            except ValueError:
                print("Введите число.")
                continue
            genre = input("Жанр: ").strip()
            if not genre:
                genre = "Неизвестно"
            watched_input = input("Статус (1-просмотрен, 0-не просмотрен): ").strip()
            watched = watched_input == "1"
            try:
                rating = int(input("Оценка (1-10): ").strip())
            except ValueError:
                rating = 0
            try:
                movie = catalog.add_movie(title, director, year, genre, watched, rating)
                print(f"Фильм добавлен с ID {movie.id}")
            except Exception as e:
                print("Ошибка:", e)
        elif choice == "2":
            if not catalog.movies:
                print("Нет фильмов.")
            else:
                for m in catalog.movies:
                    print_movie(m)
        elif choice == "3":
            movies = catalog.filter_by_watched(True)
            if not movies:
                print("Нет просмотренных фильмов.")
            else:
                for m in movies:
                    print_movie(m)
        elif choice == "4":
            movies = catalog.filter_by_watched(False)
            if not movies:
                print("Нет непросмотренных фильмов.")
            else:
                for m in movies:
                    print_movie(m)
        elif choice == "5":
            query = input("Введите часть названия или режиссёра: ").strip()
            if not query:
                print("Введите текст.")
                continue
            results = catalog.search_movies(query)
            if not results:
                print("Фильмы не найдены.")
            else:
                for m in results:
                    print_movie(m)
        elif choice == "6":
            try:
                mid = int(input("Введите ID фильма: ").strip())
            except ValueError:
                print("Некорректный ID.")
                continue
            movie = catalog.find_movie(mid)
            if not movie:
                print("Фильм не найден.")
                continue
            movie.watched = True
            print("Фильм отмечен как просмотренный.")
        elif choice == "7":
            try:
                mid = int(input("Введите ID фильма для редактирования: ").strip())
            except ValueError:
                print("Некорректный ID.")
                continue
            movie = catalog.find_movie(mid)
            if not movie:
                print("Фильм не найден.")
                continue
            print("Оставьте поле пустым, чтобы не менять.")
            new_title = input(f"Название ({movie.title}): ").strip()
            new_director = input(f"Режиссёр ({movie.director}): ").strip()
            new_year = input(f"Год ({movie.year}): ").strip()
            new_genre = input(f"Жанр ({movie.genre}): ").strip()
            new_watched = input(f"Статус (1-просмотрен, 0-не просмотрен) сейчас: {'1' if movie.watched else '0'}: ").strip()
            new_rating = input(f"Оценка ({movie.rating}): ").strip()
            updates = {}
            if new_title: updates["title"] = new_title
            if new_director: updates["director"] = new_director
            if new_year:
                try:
                    updates["year"] = int(new_year)
                except ValueError:
                    print("Год должен быть числом, пропускаем.")
            if new_genre: updates["genre"] = new_genre
            if new_watched: updates["watched"] = new_watched == "1"
            if new_rating:
                try:
                    updates["rating"] = int(new_rating)
                except ValueError:
                    print("Оценка должна быть числом, пропускаем.")
            if catalog.edit_movie(mid, **updates):
                print("Фильм обновлён.")
            else:
                print("Ошибка обновления.")
        elif choice == "8":
            try:
                mid = int(input("Введите ID фильма для удаления: ").strip())
            except ValueError:
                print("Некорректный ID.")
                continue
            if catalog.delete_movie(mid):
                print("Фильм удалён.")
            else:
                print("Фильм не найден.")
        elif choice == "9":
            stats = catalog.get_stats()
            print("\n=== СТАТИСТИКА ===")
            print(f"Всего фильмов: {stats['total']}")
            print(f"Просмотрено: {stats['watched']}")
            print(f"Не просмотрено: {stats['unwatched']}")
            print(f"Средняя оценка (только просмотренные): {stats['avg_rating']:.2f}")
            print("По жанрам:")
            for genre, count in stats['genres'].items():
                print(f"  {genre}: {count}")
        elif choice == "10":
            catalog.save_to_file()
            print("Сохранено.")
        elif choice == "11":
            catalog.load_from_file()
            print("Загружено.")
        else:
            print("Неизвестная команда.")

if __name__ == "__main__":
    main()
